data/blog/darwin-spelunking.mdx

```diff
@@ -0,0 +1,46 @@
+---
+title: 'Improving Tailscale via Apple’s open source'
+date: '2023-04-13'
+tags:
+  - 'oss'
+authors: ['mihai']
+images: ['/images/darwin-spelunking/social.png']
+summary: 'Spelunking through the Darwin kernel and userland helps Tailscale work better on macOS and iOS.'
+---
+
+# Improving Tailscale via Apple’s Open Source
+
+[Daniel Jalkut’s](https://redsweater.com/about/DanielJalkut.html) recent [Spelunking Apple’s Open Source](https://bitsplitting.org/2023/03/17/spelunking-apples-open-source/) blog post struck a very familiar chord with me. Some of my most satisfying times at my previous employer were spent [poring through WebKit internals](https://blog.persistent.info/search/label/WebKit). At Tailscale the pattern has continued, this time more focused on [Darwin](<https://en.wikipedia.org/wiki/Darwin_(operating_system)>) (the operating system at the core of macOS and iOS), its [kernel](https://en.wikipedia.org/wiki/XNU) and [userland](https://en.wikipedia.org/wiki/User_space_and_kernel_space) tools. As an example of the power of being able to peek under the hood of the operating system, here’s a recent debugging story from the Tailscale 1.36 release:
+
+We were investigating two network-interface related bugs – one where Tailscale traffic was [looping on itself](https://github.com/tailscale/tailscale/issues/5156) (instead of being passed through to the actual physical network connection) and another where Tailscale was [sending traffic over the phone’s cellular interface](https://github.com/tailscale/tailscale/issues/6565) (even though Wi-Fi was available). The [fix](https://github.com/tailscale/tailscale/pull/6566) for both of these involved more correctly “binding” Tailscale’s network requests to the active network interface (and re-binding them when it changes).
```

Comment:
What do you think about boosting the lead here up a bit? I'm mostly thinking just inverting this paragraph. Maybe something like:

The ability to peek under the hood of the operating system can be a powerful tool for debugging. “Spelunking Apple's Open Source,” a recent blog post by Daniel Jalkut, struck a very familiar chord with me, reminding me of satisfying sessions poring through WebKit internals with my previous employer. I've been able to continue that pattern at Tailscale, now more focused on Darwin (the operating system at the core of macOS and iOS), its kernel, and its userland.

Just recently, Apple's open source came in handy when debugging before the Tailscale 1.36 release. We were investigating...

Reply:
Agreed that it reads better, incorporated something along those lines.